Support complex numbers

changhc opened the pull request:
<!-- Thank you for your contribution! -->

## Change Summary

Implement a validator and a serialiser for complex numbers.

As discussed in https://github.com/pydantic/pydantic/issues/8555, since there is no official representation for complex numbers in JSON, I propose to express complex numbers as dictionaries with two keys, `real` and `imag`, with floating point values. Please find examples in the newly added test suites in this PR.

Note that function `str_as_complex` in `shared.rs` is not yet implemented as I need to discuss with the team if we want to support this. For example, we might want to accept strings like `1+2j` as a valid complex number, as long as these strings follow the format described in [python's documentation](https://docs.python.org/3/library/functions.html#complex). This implementation, however, will be a bit tricky. Using regular expression is the simplest solution, but it is going to be a bit slow since building regex is rather costly. Parsing strings using the crate `num_complex` will make the implementation very neat, but `num_complex` is more tolerant in terms of string format, which means we will need to handle strings not allowed in python and it is going to be problematic. The safest solution is probably to do exactly the same as how [cpython](https://github.com/python/cpython/blob/030b452e34bbb0096acacb70a31915b9590c8186/Objects/complexobject.c#L756) parses strings. I think that can be done in a separate PR for easier reviews.

## Related issue number

https://github.com/pydantic/pydantic/issues/8555
<!-- Are there any issues opened that will be resolved by merging this change? -->
<!-- WARNING: please use "fix #123" style references so the issue is closed when this PR is merged. -->

## Checklist

* [x] Unit tests for the changes exist
* [x] Documentation reflects the changes where applicable
* [x] Pydantic tests pass with this `pydantic-core` (except for expected changes)
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

sydney-runkle reviewed (commented):
@changhc,

Great work thus far, thanks!!

> As discussed in https://github.com/pydantic/pydantic/issues/8555, since there is no official representation for complex numbers in JSON, I propose to express complex numbers as dictionaries with two keys, real and imag, with floating point values. Please find examples in the newly added test suites in this PR.

I agree with this approach. @davidhewitt what do you think? I suppose we could have support for either this or the `1+2j` format (maybe with a config flag...)

> Note that function str_as_complex in shared.rs is not yet implemented as I need to discuss with the team if we want to support this. For example, we might want to accept strings like 1+2j as a valid complex number, as long as these strings follow the format described in [python's documentation](https://docs.python.org/3/library/functions.html#complex).

As mentioned on the `pydantic` PR, I'd be keen to support this. 

> This implementation, however, will be a bit tricky. Using regular expression is the simplest solution, but it is going to be a bit slow since building regex is rather costly. 

Perhaps we can speed things up by using the rust regex crate?

> The safest solution is probably to do exactly the same as how [cpython](https://github.com/python/cpython/blob/030b452e34bbb0096acacb70a31915b9590c8186/Objects/complexobject.c#L756) parses strings. I think that can be done in a separate PR for easier reviews.

Probably a question for @davidhewitt here.

changhc commented:
Thanks!

I've looked into the rust regex crate. It has rather limited functionality, specifically lookahead, which means we will need to handle some corner cases ourselves or create some not so straightforward logic. Another crate [fancy_regex](https://docs.rs/fancy-regex/latest/fancy_regex) has better support for regex, but introducing another dependency on a non-native library makes me a bit hesitant unless it's really necessary or beneficial enough. With regex, my biggest concern is whether or not we can capture exactly what is accepted by python. I'm concerned about accidentally accepting/rejecting some strings that are actually rejected/accepted by the complex class in python, which might bring users unexpected problems.

If the team thinks it's okay with regex (and maybe you don't feel like having this cumbersome dictionary representation at all,) I can definitely look into that.

davidhewitt commented:
What if we avoided using either a regex or `num_complex` entirely, and we used the Python logic to build `complex` instances? This is basically what we do already for `Decimal`, and it seems like a reasonable design to repeat again IMO.

changhc commented:
That sounds good to me. Then I'll work on that. On the other hand, what do
you think about the dict representation I proposed and implemented in this
PR? Should we still keep it or should we only accept strings like `1+2j`?

On Tue, 18 Jun 2024, 10:15 David Hewitt, ***@***.***> wrote:

> What if we avoided using either a regex or num_complex entirely, and we
> used the Python logic to build complex instances? This is basically what
> we do already for Decimal, and it seems like a reasonable design to
> repeat again IMO.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pydantic/pydantic-core/pull/1331#issuecomment-2175607507>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABDBC3SK3JQHP6NCYVAY5P3ZH73CHAVCNFSM6AAAAABJHEFPYSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCNZVGYYDONJQG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

sydney-runkle commented:
> That sounds good to me. Then I'll work on that. On the other hand, what do
you think about the dict representation I proposed and implemented in this
PR? Should we still keep it or should we only accept strings like `1+2j`?

We'll chat with the team in our open source sync tomorrow and get back to you :).

sydney-runkle commented:
> That sounds good to me. Then I'll work on that. On the other hand, what do
you think about the dict representation I proposed and implemented in this
PR? Should we still keep it or should we only accept strings like `1+2j`?

Let's go with the string representation for now, and if the feature lands with some popularity, we can add support for the dictionary style. For now, if folks really need that dict type serialization for complex numbers, they can use a custom serializer.

changhc commented:
I've updated the implementation. There is one test case failing with pypy, and I think that's because of some bugs in pypy. How do you usually handle discrepancy between pypy and python? Can we ignore this case for now and simply add a note in the documentation?

Another question regarding `test-pydantic-integration`. How do I handle this mutual dependency between pydantic and pydantic-core so that this test can pass?

davidhewitt commented:
> Can we ignore this case for now and simply add a note in the documentation?

Yes, we can xfail this case when running on PyPy. If you're willing to also report the case on the PyPy GitHub, it would help them be aware they need to fix :)





> Another question regarding `test-pydantic-integration`. How do I handle this mutual dependency between pydantic and pydantic-core so that this test can pass?

In this case it won't pass until we update pydantic `main` for these changes, so once we merge this PR we should make a minor release and integrate them into the main repository.

davidhewitt commented on src/errors/types.rs:
```diff
@@ -422,6 +422,9 @@ error_types! {
         whole_digits: {ctx_type: u64, ctx_fn: field_from_context},
     },
+    // Complex errors
+    ComplexParsing{},
```
```suggestion
    ComplexParsing {},
```

davidhewitt commented on src/errors/types.rs:
```diff
@@ -564,6 +567,8 @@ impl ErrorType {
             Self::DecimalMaxPlaces {..} => "Decimal input should have no more than {decimal_places} decimal place{expected_plural}",
             Self::DecimalWholeDigits {..} => "Decimal input should have no more than {whole_digits} digit{expected_plural} before the decimal point",
+            Self::ComplexParsing {..} => "Input should be a valid complex string following the rule at https://docs.python.org/3/library/functions.html#complex",
+            Self::ComplexType { .. } => "Input should be a valid complex number",
```
```suggestion
            Self::ComplexType {..} => "Input should be a valid complex number",
```

changhc replied:
Updated!

davidhewitt commented on src/input/input_python.rs:
```diff
@@ -592,6 +594,25 @@ impl<'py> Input<'py> for Bound<'py, PyAny> {
         Err(ValError::new(ErrorTypeDefaults::TimeDeltaType, self))
     }
+
+    fn validate_complex<'a>(&'a self, _py: Python<'py>) -> ValResult<ValidationMatch<EitherComplex<'py>>> {
+        if let Ok(complex) = self.downcast::<PyComplex>() {
+            return Ok(ValidationMatch::exact(EitherComplex::Py(complex.to_owned())));
+        } else if let Ok(s) = self.downcast::<PyString>() {
+            return Ok(ValidationMatch::exact(EitherComplex::Py(string_to_complex(s, self)?)));
+        } else if self.is_exact_instance_of::<PyFloat>() {
+            return Ok(ValidationMatch::exact(EitherComplex::Complex([
+                self.extract::<f64>().unwrap(),
+                0.0,
+            ])));
+        } else if self.is_exact_instance_of::<PyInt>() {
+            return Ok(ValidationMatch::exact(EitherComplex::Complex([
+                self.extract::<i64>().unwrap() as f64,
+                0.0,
+            ])));
+        }
+        Err(ValError::new(ErrorTypeDefaults::ComplexType, self))
+    }
```
We should have a think here about what behaviour we expect in `strict` mode. Probably we only want to accept `complex` objects in strict. Do we want to accept integers / floats in "strict" mode? In lax, we can accept everything for sure.

Also, similarly the `ValidationMatch::exact` calls should probably be downgraded to either `strict` or `lax` - these will set the level of priority in e.g. `complex | int` union. I think most would agree that `complex | int` would expect an integer value to _not_ be cast to a complex. We might want some according tests for some of these union cases.

changhc replied:
`strict` mode is strict in terms of typing, right? I was mostly thinking from the perspective of mathematics when I added `float` and `int`. I suppose then we should only allow `complex` objects in strict mode? Then `ValidationMatch::exact` is the same as `strict`. This sounds reasonable to me, but I'm not exactly sure about the how different `exact` and `strict` usually are.

One more question about `strict` mode: how does it apply to other input types? Is it like when the input is in JSON, we only accept strings in `strict` mode?

davidhewitt replied:
> I suppose then we should only allow `complex` objects in strict mode?

I think so, yes (note that in JSON it's less straightforward as you note). In JSON I think we only accept strings in `strict` mode, yes.

> Then `ValidationMatch::exact` is the same as `strict`.

IIRC in other cases I made subclasses `strict`, if `complex` objects allow subclassing. I'd check what I did for ints or floats.

davidhewitt commented on src/validators/complex.rs:
```diff
@@ -0,0 +1,69 @@
+}
+
+#[derive(Debug)]
+pub struct ComplexValidator {}
```
I think we _probably_ need a strict mode flag here at the very least.

What about constraints like `Le`, or `MultipleOf`? They might get painful, so I'm ok to defer those to later as long as they fail gracefully.

changhc replied:
Do we want to support constraints for complex numbers?

I thought about constraints initially when I checked other validators to get some ideas of how to implement this complex validator, but I don't think constraints are going to be useful. For example, `Le` and relevant constraints don't really make sense as complex numbers don't have a default ordering. Users can apply additional filtering ad hoc based on the ordering they define, but I don't think we can add that kind of constraints to the validator.

davidhewitt replied:
I am happy for us to not support constraints for now.

davidhewitt commented on tests/serializers/test_complex.py:
```diff
@@ -0,0 +1,39 @@
+        (complex(1, 2), '1+2j'),
+        (complex(0, 1), '1j'),
+        (complex(0, 1e-500), '0j'),
+        (complex(-float('inf'), 2), '-inf+2j'),
+        (complex(float('inf'), 2), 'inf+2j'),
+        (complex(float('nan'), 2), 'NaN+2j'),
```
It would be interesting to discuss how these interact with the `ser_json_inf_nan` setting (if at all). Maybe they shouldn't, given these are always strings.

changhc replied:
I thought about another option `allow_inf_nan` and eventually decided not to deal with it at all because I'm not really sure how this should apply to complex numbers and implementing anything with any reasoning will probably face objections from someone else.

Even though `re` and `im` parts are both `float`, the entire value is still a complex object. Rejecting something like `complex(float('inf'), 2)` kind of makes sense if we think of complex objects as "containers of float numbers", but at the same time it doesn't sound that natural because a complex number is... a complex number, a complete entity, not just "something with two float attributes, `re` and `im`".

davidhewitt replied:
I broadly agree with this, but what happens when the imaginary component is zero? That's the only case which I think could be confusing if it differs from a regular float.

davidhewitt replied:
Ah I see, we get a `+0j` part still. So yes let's not worry about that setting.

davidhewitt reviewed (commented):
Thanks, this is looking great! There's a few edge cases to discuss wrt config, constraints, and unions. Beyond that, this implementation looks rock-solid 🪨

changhc commented:
Hi @davidhewitt, how should we proceed with this? Do we wait for input from the team or should we manage to make a decision here ourselves?

sydney-runkle commented:
Hey @changhc,

Sorry we've dropped the ball on feedback here. Will chat with the team early next week so that we can move this forward and include it in our v2.9 release 🚀 !

changhc commented:
Thanks @sydney-runkle! I'm making some changes to address David's comments. They will be ready soon!

changhc commented:
Hi @davidhewitt, I've implemented the strict mode for complex numbers. As we discussed, when strict mode is on,
* only complex objects are accepted for python input.
* only complex strings are accepted for JSON input.

Because of the strict mode, I tweaked the messages for validation errors a little bit to make errors less confusing to users. Specifically for python input,
* when strict mode is on, tell users that only python complex objects are accepted.
* when strict mode is off and the input value is an invalid complex string, give the generic message explaining all accepted input values instead of telling users to correct the string since at this point we are not sure if users actually want to place a complex string or just give a value of an incorrect type.

For other input types, the error is rather simple as complex strings are the only accepted input values.

I also added some test cases for dictionaries involving complex keys just to make sure things work as expected.

davidhewitt commented on src/input/input_python.rs:
```diff
@@ -598,6 +600,40 @@ impl<'py> Input<'py> for Bound<'py, PyAny> {
+            return Ok(ValidationMatch::strict(EitherComplex::Py(complex.to_owned())));
+        }
+        if strict {
+            return Err(ValError::new(ErrorTypeDefaults::ComplexTypePyStrict, self));
```
I think rather than a new type here, we can use isinstance error:

```suggestion
            return Err(ValError::new(
                ErrorType::IsInstanceOf {
                    class: PyComplex::type_object(py).qualname()
                        .and_then(|name| name.extract())
                        .unwrap_or_else(|_| "complex".to_owned()),
                    context: None,
                },
                self,
            ))
```

changhc replied:
Updated. The error message now looks like:
```
pydantic_core._pydantic_core.ValidationError: 1 validation error for complex
  Input should be an instance of complex [type=is_instance_of, input_value='foobar', input_type=str]
  For further information visit https://errors.pydantic.dev/2.8/v/is_instance_of 
```

davidhewitt reviewed (commented):
This generally looks great to me, thanks.

The only change I'd like to see would be to remove the custom `py_strict` error type and use an instance-of error instead.

sydney-runkle commented:
@davidhewitt, are we ready to move this across the line?

davidhewitt commented:
Yep, LGTM! Thanks @changhc

changhc commented:
Thanks! I'll update https://github.com/pydantic/pydantic/pull/9654 once this PR is merged and the next small release includes these changes.

sydney-runkle commented:
Great! I'll get this merged tomorrow 🚀

sydney-runkle commented:
Great work @changhc, going to go ahead and merge this - the failing integration tests should soon be fixed with your branch.

Do we need to make any more updates to https://github.com/pydantic/pydantic/pull/9654 other than supporting a new version of `pydantic-core` with this change?
